File: agent-blackbox/upmarto-sdk-ts/src/bootstrap.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";

import { writeProjectConfig } from "./config.js";

export const NOT_CONFIGURED_MSG =
  "Upmarto backend not configured. Run: upmarto init";

const PROBE_TIMEOUT_MS = 400;
const SCAN_START = 59_000;
const SCAN_END = 60_000;
const SCAN_BATCH = 64;

export interface RuntimeFile {
  api_url: string;
  port: number;
  detected_at?: string;
}

export function runtimeConfigPath(workspace: string): string {
  return join(resolve(workspace), ".upmarto", "runtime.json");
}

export function readDotenvPort(workspace: string): number | null {
  const envPath = join(resolve(workspace), ".env");
  if (!existsSync(envPath)) return null;
  for (const line of readFileSync(envPath, "utf8").split("\n")) {
    const trimmed = line.trim();
    if (trimmed.startsWith("APP_PORT=")) {
      const port = Number.parseInt(
        trimmed.slice("APP_PORT=".length).trim().replace(/^"|"$/g, ""),
        10,
      );
      return Number.isFinite(port) ? port : null;
    }
  }
  return null;
}

export async function probeBackend(apiUrl: string): Promise<boolean> {
  const url = `${apiUrl.replace(/\/$/, "")}/config`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), PROBE_TIMEOUT_MS);
  try {
    const res = await fetch(url, { signal: controller.signal });
    if (!res.ok) return false;
    const body = (await res.json()) as { api_version?: string };
    return body.api_version === "v1";
  } catch {
    return false;
  } finally {
    clearTimeout(timer);
  }
}

export function readRuntimeFile(workspace: string): RuntimeFile | null {
  const path = runtimeConfigPath(workspace);
  if (!existsSync(path)) return null;
  try {
    return JSON.parse(readFileSync(path, "utf8")) as RuntimeFile;
  } catch {
    return null;
  }
}

export function writeRuntimeFile(
  workspace: string,
  apiUrl: string,
  port: number,
): void {
  const dir = join(resolve(workspace), ".upmarto");
  mkdirSync(dir, { recursive: true });
  const runtime: RuntimeFile = {
    api_url: apiUrl.replace(/\/$/, ""),
    port,
    detected_at: new Date().toISOString(),
  };
  writeFileSync(runtimeConfigPath(workspace), JSON.stringify(runtime, null, 2), "utf8");
}

function parsePort(apiUrl: string): number {
  const match = apiUrl.match(/:(\d+)\/?$/);
  return match ? Number.parseInt(match[1], 10) : 0;
}

async function scanLocalPorts(): Promise<string | null> {
  const ports: number[] = [];
  for (let p = SCAN_START; p <= SCAN_END; p++) ports.push(p);

  for (let i = 0; i < ports.length; i += SCAN_BATCH) {
    const chunk = ports.slice(i, i + SCAN_BATCH);
    const results = await Promise.all(
      chunk.map(async (port) => {
        const url = `http://127.0.0.1:${port}`;
        return (await probeBackend(url)) ? url : null;
      }),
    );
    const hit = results.find((r) => r !== null);
    if (hit) return hit;
  }
  return null;
}

export async function discoverBackend(workspace: string): Promise<string> {
  const runtime = readRuntimeFile(workspace);
  if (runtime?.api_url && (await probeBackend(runtime.api_url))) {
    return runtime.api_url.replace(/\/$/, "");
  }

  const envPort = readDotenvPort(workspace);
  if (envPort) {
    const url = `http://127.0.0.1:${envPort}`;
    if (await probeBackend(url)) return url;
  }

  const scanned = await scanLocalPorts();
  if (scanned) return scanned;

  throw new Error(
    `No running Upmarto backend found on 127.0.0.1:${SCAN_START}-${SCAN_END}. ` +
      "Start the server (`cargo run`) then run: upmarto init",
  );
}

export async function validateWorkspaceAccess(
  workspace: string,
  apiUrl: string,
): Promise<void> {
  if (!apiUrl) {
    throw new Error(NOT_CONFIGURED_MSG);
  }

  const envOverride = Boolean(
    process.env.UPMARTO_URL?.trim() || process.env.BLACKBOX_URL?.trim(),
  );
  const projectConfig = join(resolve(workspace), ".upmarto", "config.json");

  if (!envOverride && !existsSync(projectConfig)) {
    throw new Error(NOT_CONFIGURED_MSG);
  }

  if (!envOverride && !(await probeBackend(apiUrl))) {
    throw new Error(
      `Upmarto backend not reachable at ${apiUrl}. Run: upmarto init`,
    );
  }
}

export async function bootstrapWorkspace(
  workspace: string,
  explicitApiUrl?: string,
): Promise<string> {
  let apiUrl = explicitApiUrl?.replace(/\/$/, "");
  if (!apiUrl) {
    const env = process.env.UPMARTO_URL?.trim();
    apiUrl = env ? env.replace(/\/$/, "") : await discoverBackend(workspace);
  }

  if (!(await probeBackend(apiUrl))) {
    throw new Error(
      `Upmarto backend not reachable at ${apiUrl}. ` +
        "Start the server (`cargo run`) then run: upmarto init",
    );
  }

  writeProjectConfig(workspace, {
    api_url: apiUrl,
    project_id: "auto",
    auto_capture: true,
    batch_size: 50,
    flush_interval_ms: 2000,
    retry_max: 5,
  });
  writeRuntimeFile(workspace, apiUrl, parsePort(apiUrl));
  process.env.UPMARTO_URL = apiUrl;
  return apiUrl;
}

File: agent-blackbox/upmarto-sdk-rust/src/bootstrap.rs
use std::path::Path;
use std::time::Duration;

use serde::{Deserialize, Serialize};

use crate::config::{project_config_path, read_dotenv_port, runtime_config_path};
use crate::SdkError;

pub const NOT_CONFIGURED_MSG: &str =
    "Upmarto backend not configured. Run: upmarto init";

const PROBE_TIMEOUT: Duration = Duration::from_millis(400);
const SCAN_PORT_START: u16 = 59_000;
const SCAN_PORT_END: u16 = 60_000;
const SCAN_BATCH: usize = 64;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeFile {
    pub api_url: String,
    pub port: u16,
    #[serde(default)]
    pub detected_at: String,
}

/// GET /config — returns true when an Upmarto v1 backend responds.
pub async fn probe_backend(api_url: &str) -> bool {
    let url = format!("{}/config", api_url.trim_end_matches('/'));
    let client = match reqwest::Client::builder().timeout(PROBE_TIMEOUT).build() {
        Ok(c) => c,
        Err(_) => return false,
    };

    let Ok(res) = client.get(&url).send().await else {
        return false;
    };
    if !res.status().is_success() {
        return false;
    }

    let Ok(body) = res.json::<serde_json::Value>().await else {
        return false;
    };

    body.get("api_version")
        .and_then(|v| v.as_str())
        .is_some_and(|v| v == "v1")
}

pub fn read_runtime_file(workspace: &Path) -> Option<RuntimeFile> {
    let raw = std::fs::read_to_string(runtime_config_path(workspace)).ok()?;
    serde_json::from_str(&raw).ok()
}

pub fn write_runtime_file(workspace: &Path, api_url: &str, port: u16) -> std::io::Result<()> {
    let dir = workspace.join(".upmarto");
    std::fs::create_dir_all(&dir)?;
    let runtime = RuntimeFile {
        api_url: api_url.trim_end_matches('/').to_string(),
        port,
        detected_at: chrono::Local::now().to_rfc3339(),
    };
    let json = serde_json::to_string_pretty(&runtime)?;
    std::fs::write(runtime_config_path(workspace), json)
}

fn parse_port_from_url(api_url: &str) -> u16 {
    api_url
        .trim_end_matches('/')
        .rsplit(':')
        .next()
        .and_then(|p| p.parse().ok())
        .unwrap_or(0)
}

/// Discover a live Upmarto backend (runtime.json → .env APP_PORT → port scan).
pub async fn discover_backend(workspace: &Path) -> Result<String, SdkError> {
    if let Some(runtime) = read_runtime_file(workspace) {
        if probe_backend(&runtime.api_url).await {
            return Ok(runtime.api_url);
        }
    }

    if let Some(port) = read_dotenv_port(workspace) {
        let url = format!("http://127.0.0.1:{port}");
        if probe_backend(&url).await {
            return Ok(url);
        }
    }

    if let Some(url) = scan_local_ports().await {
        return Ok(url);
    }

    Err(SdkError::Config(format!(
        "No running Upmarto backend found on 127.0.0.1:{SCAN_PORT_START}-{SCAN_PORT_END}. \
         Start the server (`cargo run`) then run: upmarto init"
    )))
}

async fn scan_local_ports() -> Option<String> {
    let ports: Vec<u16> = (SCAN_PORT_START..=SCAN_PORT_END).collect();

    for chunk in ports.chunks(SCAN_BATCH) {
        let mut tasks = Vec::with_capacity(chunk.len());
        for &port in chunk {
            tasks.push(tokio::spawn(async move {
                let url = format!("http://127.0.0.1:{port}");
                if probe_backend(&url).await {
                    Some(url)
                } else {
                    None
                }
            }));
        }

        for task in tasks {
            if let Ok(Some(url)) = task.await {
                return Some(url);
            }
        }
    }

    None
}

/// Validate workspace config before SDK use. Skips reachability probe when `UPMARTO_URL` is set.
pub async fn validate_workspace_access(workspace: &Path, api_url: &str) -> Result<(), SdkError> {
    if api_url.is_empty() {
        return Err(SdkError::Config(NOT_CONFIGURED_MSG.into()));
    }

    let env_override = std::env::var("UPMARTO_URL")
        .ok()
        .is_some_and(|v| !v.trim().is_empty());

    if !env_override && !project_config_path(workspace).exists() {
        return Err(SdkError::Config(NOT_CONFIGURED_MSG.into()));
    }

    if !env_override && !probe_backend(api_url).await {
        return Err(SdkError::Config(format!(
            "Upmarto backend not reachable at {api_url}. Run: upmarto init"
        )));
    }

    Ok(())
}

/// Bootstrap: discover, probe, write config + runtime.json.
pub async fn bootstrap_workspace(
    workspace: &Path,
    explicit_api_url: Option<String>,
    cfg: &crate::config::UpmartoConfig,
) -> Result<String, SdkError> {
    let api_url = match explicit_api_url {
        Some(url) => url.trim_end_matches('/').to_string(),
        None => std::env::var("UPMARTO_URL")
            .ok()
            .filter(|u| !u.trim().is_empty())
            .map(|u| u.trim_end_matches('/').to_string())
            .unwrap_or(discover_backend(workspace).await?),
    };

    if !probe_backend(&api_url).await {
        return Err(SdkError::Config(format!(
            "Upmarto backend not reachable at {api_url}. \
             Start the server (`cargo run`) then run: upmarto init"
        )));
    }

    let mut final_cfg = cfg.clone();
    final_cfg.api_url = api_url.clone();
    crate::config::write_project_config(workspace, &final_cfg)?;
    let port = parse_port_from_url(&api_url);
    write_runtime_file(workspace, &api_url, port)?;

    std::env::set_var("UPMARTO_URL", &api_url);

    Ok(api_url)
}
